### PR TITLE
DD-183 DD-250 Use dans-scala-lib and sort Tasks

### PR DESCRIPTION
**Pull request by ChrisMuller80**
Fixes [DD-250](https://drivenbydata.atlassian.net/browse/DD-250) and [DD-183](https://drivenbydata.atlassian.net/browse/DD-183)

# Description of changes

This PR combines two issues because they are closely related.  
The changes are: 
A. Use `TaskQueue` classes from `dans-scala-lib`
B. Add sorting of DepositIngestTasks using the abovementioned classes

# How to test
Adding multiple versions into Dataverse is not yet implemented. 
To test run tests in `src/test/scala/nl.knaw.dans.easy.dd2d/DepositIngestTaskSortSpec.scala`

# Related PRs 
* See links above

# Notify
@DANS-KNAW/dataversedans

### CODE REVIEW

File: src/main/scala/nl.knaw.dans.easy.dd2d/DepositSorter.scala [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+import java.time.LocalDateTime
+import java.time.format.DateTimeFormatter
+import java.util.Properties
+
+import nl.knaw.dans.lib.logging.DebugEnhancedLogging
+import nl.knaw.dans.lib.taskqueue.{ Task, TaskSorter }
+
+class DepositSorter extends TaskSorter[Deposit] with DebugEnhancedLogging {
+
+  private case class DepositsSortInfo(name: String, timeStamp: LocalDateTime, isVersionOf: Option[String], depositTask: DepositIngestTask)
```
Comment (janvanmansum) [inline src/main/scala/nl.knaw.dans.easy.dd2d/DepositSorter.scala:28]: Personally, I find the first to typedefs below not really worthwhile, but you can keep them in if you do. The reason why I wouldn't is that List[X] already reads naturally as "XList". On the other hand VersionMap does add extra semantic info.

File: src/test/scala/nl.knaw.dans.easy.dd2d/queue/TriggerTask.scala [hunk 2]
```diff
@@ -15,15 +15,19 @@
...
 
-case class TriggerTask() extends Task with DebugEnhancedLogging {
+case class TriggerTask() extends Task[Deposit] with DebugEnhancedLogging {
   var triggered = false
 
   override def run(): Try[Unit] = Try {
     trace(())
     triggered = true
   }
+
```
Comment (jo-pol) [inline src/test/scala/nl.knaw.dans.easy.dd2d/queue/TriggerTask.scala:31]: I wonder if we can't use mocking above

File: src/test/scala/nl.knaw.dans.easy.dd2d/queue/ActiveTaskQueueSpec.scala [hunk 3]
```diff
@@ -15,13 +15,16 @@
...
 package nl.knaw.dans.easy.dd2d.queue
 
+import nl.knaw.dans.easy.dd2d.Deposit
+import nl.knaw.dans.lib.taskqueue.ActiveTaskQueue
 import org.scalatest.{ FlatSpec, Matchers, OneInstancePerTest }
 
 class ActiveTaskQueueSpec extends FlatSpec with Matchers with OneInstancePerTest {
 
   // TODO: how to make this test robust?
+  //TODO: should this be moved to dans-scala-lib?
```
Comment (janvanmansum) [inline src/test/scala/nl.knaw.dans.easy.dd2d/queue/ActiveTaskQueueSpec.scala:25]: Answer: definitely! Please remove it from the project. I will add it to dans-scala-lib.

File: src/main/scala/nl.knaw.dans.easy.dd2d/DepositSorter.scala [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+  private def sortVersions(groupedDeposits: VersionMap): DepositIngestTaskList = {
+    groupedDeposits.mapValues(k => sortByTimestamp(k))
+      .flatMap(_._2)
+      .map(_.depositTask)
+      .toList
+  }
+
+  private def removeDatasetsWithoutFirstVersion(firstVersions: VersionMap, laterVersions: VersionMap): VersionMap = {
+    val datasetsWithoutFirstVersions = for (k <- laterVersions if !firstVersions.keySet.contains(k._1)) yield k
+    datasetsWithoutFirstVersions.foreach(v => logger.error(s"No first version was found for dataset ${ v._1 }. The dataset was not imported into Dataverse"))
```
Comment (jo-pol) [inline src/main/scala/nl.knaw.dans.easy.dd2d/DepositSorter.scala:92]: in general, avoid `._1` and `._2` and find something more readable  
```suggestion
    datasetsWithoutFirstVersions.foreach{ case (k,v) => 
      logger.error(s"No first version was found for dataset $ v. The dataset was not imported into Dataverse")
    }
```

File: src/main/scala/nl.knaw.dans.easy.dd2d/DepositSorter.scala [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+  override def sort(tasks: List[Task[Deposit]]): List[Task[Deposit]] = {
+    val sortInfoList = tasks.map(d => getDepositSortInfo(d.asInstanceOf[DepositIngestTask]))
+    val groupedDeposits = groupVersions(sortInfoList)
+    sortVersions(groupedDeposits)
+  }
+
+  /**
+   * creates a DepositsSortInfo object that contains necessary sorting information
+   *
+   * @param depositIngestTask
```
Comment (janvanmansum) [inline src/main/scala/nl.knaw.dans.easy.dd2d/DepositSorter.scala:51]: Missing parameter description (says IntelliJ...) You may also leave out the scaladoc as this is a private def.

File: src/main/scala/nl.knaw.dans.easy.dd2d/DepositSorter.scala [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+
+  private def parseTimestamp(timestampString: String): LocalDateTime = {
+    LocalDateTime.parse(timestampString, DateTimeFormatter.ISO_OFFSET_DATE_TIME)
+  }
+
+  private def sortByTimestamp(sortInfoList: SortInfoList): SortInfoList = {
+    sortInfoList.sortBy(_.timeStamp)
+  }
+
+  implicit def ordered: Ordering[LocalDateTime] = (x: LocalDateTime, y: LocalDateTime) => x compareTo y
```
Comment (janvanmansum) [inline src/main/scala/nl.knaw.dans.easy.dd2d/DepositSorter.scala:105]: Move this to the first line of sortByTimestamp, as that is the only scope where this is relevant.

File: src/main/scala/nl.knaw.dans.easy.dd2d/DepositSorter.scala [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+
+    val laterVersionsUpdated = removeDatasetsWithoutFirstVersion(firstVersions, laterVersions)
+
+    (firstVersions.keySet ++ laterVersionsUpdated.keySet)
+      .map(k => k -> (firstVersions(k) ++ laterVersionsUpdated(k)))
+      .toMap
+  }
+
+  private def sortVersions(groupedDeposits: VersionMap): DepositIngestTaskList = {
+    groupedDeposits.mapValues(k => sortByTimestamp(k))
```
Comment (jo-pol) [inline src/main/scala/nl.knaw.dans.easy.dd2d/DepositSorter.scala:84]: ALT-enter at first `k` suggests
```suggestion
    groupedDeposits.mapValues(sortByTimestamp(_))
```
then you get an orange warning  on (_), ALT-enter again and you'l get
```suggestion
    groupedDeposits.mapValues(sortByTimestamp)
```
What indeed reads better than
```suggestion
    groupedDeposits.mapValues(_.sortBy(_.timeStamp))
```

File: src/main/scala/nl.knaw.dans.easy.dd2d/DepositSorter.scala [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+
+import nl.knaw.dans.lib.logging.DebugEnhancedLogging
+import nl.knaw.dans.lib.taskqueue.{ Task, TaskSorter }
+
+class DepositSorter extends TaskSorter[Deposit] with DebugEnhancedLogging {
+
+  private case class DepositsSortInfo(name: String, timeStamp: LocalDateTime, isVersionOf: Option[String], depositTask: DepositIngestTask)
+  private type DepositIngestTaskList = List[DepositIngestTask]
+  private type SortInfoList = List[DepositsSortInfo]
+  private type VersionMap = Map[String, List[DepositsSortInfo]]
```
Comment (janvanmansum) [inline src/main/scala/nl.knaw.dans.easy.dd2d/DepositSorter.scala:31]: UpdateDepositsMap ?  It maps from a base deposit to the list of update-deposits for that base deposits, right? It took me a while to understand that, so I am trying to find an alternative name of this type that captures its meaning better.

File: src/main/scala/nl.knaw.dans.easy.dd2d/DepositSorter.scala [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+
+  /**
+   * sorts the (DepositIngest)Tasks by dataset and per dataset by version
+   *
+   * @param tasks list of DepositIngestTasks
+   * @return the sorted list of deposit directories
+   */
+  override def sort(tasks: List[Task[Deposit]]): List[Task[Deposit]] = {
+    val sortInfoList = tasks.map(d => getDepositSortInfo(d.asInstanceOf[DepositIngestTask]))
+    val groupedDeposits = groupVersions(sortInfoList)
```
Comment (janvanmansum) [inline src/main/scala/nl.knaw.dans.easy.dd2d/DepositSorter.scala:44]: depositSequences = getDepositSequences(sortInfoList)

File: src/main/scala/nl.knaw.dans.easy.dd2d/DepositSorter.scala [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+    sortVersions(groupedDeposits)
```
Comment (janvanmansum) [inline src/main/scala/nl.knaw.dans.easy.dd2d/DepositSorter.scala:45]: sortByVersion

File: src/test/scala/nl.knaw.dans.easy.dd2d/DepositIngestTaskSortSpec.scala [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+  val DATASETVERSIONS_WITHOUT_FIRST_VERSION = 1
+
+  "The result" should "be a list with size" + (depositDirs.size - DATASETVERSIONS_WITHOUT_FIRST_VERSION) in {
+    val result = depositSorter.sort(taskList)
+    result should have size depositDirs.size - DATASETVERSIONS_WITHOUT_FIRST_VERSION
+  }
+
+  it should "be a correctly ordered list" in {
+    val result = depositSorter.sort(taskList)
+    result.map(_.getTarget.dir.name) should be(List("deposit3_first", "deposit3_a", "deposit3_b", "deposit1_first", "deposit1_a", "deposit1_b", "deposit2_first", "deposit2_a"))
```
Comment (janvanmansum) [inline src/test/scala/nl.knaw.dans.easy.dd2d/DepositIngestTaskSortSpec.scala:41]: How do we know that the depost1_* sequence will not be first? We don't care about that, so the test should not depend on it either.

**Review by janvanmansum — CHANGES_REQUESTED** [review]
Looks good! Still, I have added some comments. Some are more important than others. The naming the the sorter is important to improve, to make your algortihm more readable. I have given some suggestions, but we can talk about alternatives.

File: src/main/scala/nl.knaw.dans.easy.dd2d/DepositSorter.scala [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+    val laterVersionsUpdated = removeDatasetsWithoutFirstVersion(firstVersions, laterVersions)
+
+    (firstVersions.keySet ++ laterVersionsUpdated.keySet)
+      .map(k => k -> (firstVersions(k) ++ laterVersionsUpdated(k)))
+      .toMap
+  }
+
+  private def sortVersions(groupedDeposits: VersionMap): DepositIngestTaskList = {
+    groupedDeposits.mapValues(k => sortByTimestamp(k))
+      .flatMap(_._2)
```
[the remark this answers is not in the document]
Comment (ChrisMuller80) [reply src/main/scala/nl.knaw.dans.easy.dd2d/DepositSorter.scala:85]: mapValues is a function that transforms the map by applying a function to every value in the map which is what is needed here.

File: src/main/scala/nl.knaw.dans.easy.dd2d/DepositSorter.scala [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+  private val CREATED = "Created"
+
+  /**
+   * sorts the (DepositIngest)Tasks by dataset and per dataset by version
+   *
+   * @param tasks list of DepositIngestTasks
+   * @return the sorted list of deposit directories
+   */
+  override def sort(tasks: List[Task[Deposit]]): List[Task[Deposit]] = {
+    val sortInfoList = tasks.map(d => getDepositSortInfo(d.asInstanceOf[DepositIngestTask]))
```
[the remark this answers is not in the document]
Comment (ChrisMuller80) [reply src/main/scala/nl.knaw.dans.easy.dd2d/DepositSorter.scala:43]: I tried it, doesn't work.

File: src/main/scala/nl.knaw.dans.easy.dd2d/DepositIngestTask.scala [hunk 14]
```diff
@@ -107,4 +107,8 @@ case class DepositIngestTask(deposit: Deposit, dansBagValidator: DansBagValidato
   private def publishDataset(datasetId: String): Try[Unit] = {
     dataverse.dataset(datasetId, isPersistentId = true).publish("major").map(_ => ())
   }
+
+  override def getTarget: Deposit = {
+    deposit
+  }
```
[the remark this answers is not in the document]
Comment (ChrisMuller80) [reply src/main/scala/nl.knaw.dans.easy.dd2d/DepositIngestTask.scala:113]: To be changed in other PR about using `dans-dataverse-scala-lib`

File: src/test/scala/nl.knaw.dans.easy.dd2d/queue/TriggerTask.scala (hunk 2 above)
Comment (ChrisMuller80) [reply src/test/scala/nl.knaw.dans.easy.dd2d/queue/TriggerTask.scala:31]: To be moved to external library

File: src/main/scala/nl.knaw.dans.easy.dd2d/Inbox.scala [hunk 15]
```diff
@@ -15,55 +15,31 @@
...
-      d => {
-        debug(s"Adding $d")
-        q.add(DepositIngestTask(Deposit(d), dansBagValidator, dataverse, publish = autoPublish))
-      }
+  override def createTask(f: File): Option[DepositIngestTask] = {
+    f match {
+      case f if isDeposit(f) => Some(DepositIngestTask(Deposit(f), dansBagValidator, dataverse, autoPublish))
+      case _ => None
     }
   }
```
[the remark this answers is not in the document]
Comment (ChrisMuller80) [reply src/main/scala/nl.knaw.dans.easy.dd2d/Inbox.scala:41]: done

File: src/main/scala/nl.knaw.dans.easy.dd2d/DepositSorter.scala (hunk 8 above)
Comment (ChrisMuller80) [reply src/main/scala/nl.knaw.dans.easy.dd2d/DepositSorter.scala:31]: More or less.
The key is the name of the first version of the Dataset. The value is a List[DepositSortinfo] in which **all** the ordered Deposit versions of that Dataset can be found.

File: src/main/scala/nl.knaw.dans.easy.dd2d/Inbox.scala (hunk 15 above)
[the remark this answers is not in the document]
Comment (ChrisMuller80) [reply src/main/scala/nl.knaw.dans.easy.dd2d/Inbox.scala:41]: Not implemented yet

File: src/main/scala/nl.knaw.dans.easy.dd2d/DepositSorter.scala (hunk 13 above)
[the remark this answers is not in the document]
Comment (janvanmansum) [reply src/main/scala/nl.knaw.dans.easy.dd2d/DepositSorter.scala:45]: I think Scala doesn't like this contruct if the _ is in a nested argument.

**Review by janvanmansum — APPROVED** [review]

**Pull request merged**